src/Microsoft.AspNetCore.ResponseCaching/Internal/ResponseCachePolicyProvider.cs

```diff
@@ -158,44 +186,69 @@ public virtual bool IsResponseCacheable(ResponseCacheContext context)
-                context.Logger.LogExpirationMinFreshAdded(minFresh.Value);
+                var minFresh = TimeSpan.FromSeconds(seconds);
+                age += minFresh;
+                context.Logger.LogExpirationMinFreshAdded(minFresh);
             }
 
             // Validate shared max age, this overrides any max age settings for shared caches
-            var sharedMaxAge = cachedControlHeaders.SharedMaxAge;
-            if (age >= sharedMaxAge)
+            TimeSpan? cachedSharedMaxAge = null;
+            if (ParsingHelpers.TryGetHeaderValue(cachedControlHeaders, CacheControlValues.SharedMaxAgeString, out seconds))
+            {
+                cachedSharedMaxAge = TimeSpan.FromSeconds(seconds);
+            }
+
+            if (age >= cachedSharedMaxAge)
             {
                 // shared max age implies must revalidate
-                context.Logger.LogExpirationSharedMaxAgeExceeded(age, sharedMaxAge.Value);
+                context.Logger.LogExpirationSharedMaxAgeExceeded(age, cachedSharedMaxAge.Value);
                 return false;
             }
-            else if (!sharedMaxAge.HasValue)
+            else if (!cachedSharedMaxAge.HasValue)
             {
-                var cachedMaxAge = cachedControlHeaders.MaxAge;
-                var requestMaxAge = context.RequestCacheControlHeaderValue.MaxAge;
+                TimeSpan? requestMaxAge = null;
+                if (ParsingHelpers.TryGetHeaderValue(requestCacheControlHeaders, CacheControlValues.MaxAgeString, out seconds))
+                {
+                    requestMaxAge = TimeSpan.FromSeconds(seconds);
+                }
+
+                TimeSpan? cachedMaxAge = null;
+                if (ParsingHelpers.TryGetHeaderValue(cachedControlHeaders, CacheControlValues.MaxAgeString, out seconds))
+                {
+                    cachedMaxAge = TimeSpan.FromSeconds(seconds);
+                }
+
                 var lowestMaxAge = cachedMaxAge < requestMaxAge ? cachedMaxAge : requestMaxAge ?? cachedMaxAge;
                 // Validate max age
                 if (age >= lowestMaxAge)
                 {
                     // Must revalidate
-                    if (cachedControlHeaders.MustRevalidate)
+                    foreach (var header in cachedControlHeaders)
                     {
-                        context.Logger.LogExpirationMustRevalidate(age, lowestMaxAge.Value);
-                        return false;
+                        if (header.IndexOf(CacheControlValues.MustRevalidateString, StringComparison.OrdinalIgnoreCase) != -1)
+                        {
+                            context.Logger.LogExpirationMustRevalidate(age, lowestMaxAge.Value);
+                            return false;
+                        }
+                    }
+
+                    TimeSpan? requestMaxStale = null;
+                    if (ParsingHelpers.TryGetHeaderValue(requestCacheControlHeaders, CacheControlValues.MaxStaleString, out seconds))
```

Comment:
Actually this can be simplified by having a `ParsingHelpers.TryParseTimeSpan` (or some consistent naming) that returns a TimeSpan.